File: sdk/python/moondb/graph.py
"""Graph engine helpers for MoonDB (GRAPH.* commands).

Wraps GRAPH.CREATE, GRAPH.ADDNODE, GRAPH.ADDEDGE, GRAPH.DELETE,
GRAPH.NEIGHBORS, GRAPH.QUERY, GRAPH.RO_QUERY, GRAPH.EXPLAIN,
GRAPH.INFO, GRAPH.LIST, GRAPH.VSEARCH, GRAPH.HYBRID, GRAPH.PROFILE.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Sequence, Union

from .types import GraphEdge, GraphNode, QueryResult, encode_vector


class GraphCommands:
    """Graph command helpers.

    Not instantiated directly -- accessed via ``MoonClient.graph``.

    Example::

        from moondb import MoonClient
        client = MoonClient()
        client.graph.create("social")
        node_id = client.graph.add_node("social", "Person", name="Alice", age="30")
    """

    def __init__(self, client: Any) -> None:
        self._client = client

    def create(self, name: str) -> str:
        """Create a named graph.

        Args:
            name: Graph name.

        Returns:
            "OK" on success.

        Example::

            client.graph.create("knowledge")
        """
        return self._client.execute_command("GRAPH.CREATE", name)  # type: ignore[no-any-return]

    def delete(self, name: str) -> str:
        """Delete a named graph and all its nodes/edges.

        Args:
            name: Graph name.

        Returns:
            "OK" on success.

        Example::

            client.graph.delete("old_graph")
        """
        return self._client.execute_command("GRAPH.DELETE", name)  # type: ignore[no-any-return]

    def add_node(
        self,
        graph: str,
        label: str,
        *,
        vector_field: Optional[str] = None,
        vector: Optional[Union[Sequence[float], bytes]] = None,
        **properties: str,
    ) -> int:
        """Add a node to the graph.

        Args:
            graph: Graph name.
            label: Node label (e.g. "Person", "Document").
            vector_field: Optional vector field name for vector-enabled nodes.
            vector: Optional vector data (list of floats or bytes).
            **properties: Key-value properties to store on the node.

        Returns:
            The numeric node ID assigned by the server.

        Example::

            node_id = client.graph.add_node(
                "social", "Person", name="Alice", age="30"
            )
            # With vector:
            node_id = client.graph.add_node(
                "knowledge", "Document",
                vector_field="embedding", vector=[0.1, 0.2, ...],
                title="My Doc",
            )
        """
        args: list[Any] = ["GRAPH.ADDNODE", graph, label]

        for k, v in properties.items():
            args.extend([k, v])

        if vector_field and vector is not None:
            blob = vector if isinstance(vector, bytes) else encode_vector(vector)
            args.extend(["VECTOR", vector_field, blob])

        result = self._client.execute_command(*args)
        return int(result)

    def add_edge(
        self,
        graph: str,
        src_id: int,
        dst_id: int,
        edge_type: str,
        *,
        weight: Optional[float] = None,
        **properties: str,
    ) -> str:
        """Add an edge between two nodes.

        Args:
            graph: Graph name.
            src_id: Source node ID.
            dst_id: Destination node ID.
            edge_type: Relationship type (e.g. "KNOWS", "LINKS_TO").
            weight: Optional edge weight.
            **properties: Key-value properties on the edge.

        Returns:
            "OK" on success.

        Example::

            client.graph.add_edge("social", 1, 2, "KNOWS", weight=0.9)
        """
        args: list[Any] = ["GRAPH.ADDEDGE", graph, str(src_id), str(dst_id), edge_type]

        for k, v in properties.items():
            args.extend([k, v])

        if weight is not None:
            args.extend(["WEIGHT", str(weight)])

        return self._client.execute_command(*args)  # type: ignore[no-any-return]

    def neighbors(
        self,
        graph: str,
        node_id: int,
        *,
        edge_type: Optional[str] = None,
        depth: int = 1,
    ) -> List[GraphNode]:
        """Get neighbors of a node.

        Args:
            graph: Graph name.
            node_id: Node ID to query.
            edge_type: Optional edge type filter.
            depth: Number of hops (default 1).

        Returns:
            List of GraphNode representing neighbors.

        Example::

            neighbors = client.graph.neighbors("social", 1, depth=2)
            for n in neighbors:
                print(n.node_id, n.label, n.properties)
        """
        args: list[Any] = ["GRAPH.NEIGHBORS", graph, str(node_id)]
        if edge_type:
            args.extend(["TYPE", edge_type])
        if depth > 1:
            args.extend(["DEPTH", str(depth)])

        raw = self._client.execute_command(*args)
        return _parse_neighbors(raw)

    def query(self, graph: str, cypher: str) -> QueryResult:
        """Execute a Cypher query (read-write).

        Args:
            graph: Graph name.
            cypher: Cypher query string.

        Returns:
            QueryResult with headers, rows, and stats.

        Example::

            result = client.graph.query(
                "social",
                "CREATE (a:Person {name: 'Alice'})-[:KNOWS]->(b:Person {name: 'Bob'}) RETURN a, b"
            )
        """
        raw = self._client.execute_command("GRAPH.QUERY", graph, cypher)
        return _parse_query_result(raw)

    def ro_query(self, graph: str, cypher: str) -> QueryResult:
        """Execute a read-only Cypher query (rejects writes).

        Args:
            graph: Graph name.
            cypher: Cypher query string (must not contain CREATE/DELETE/SET/MERGE).

        Returns:
            QueryResult with headers, rows, and stats.

        Example::

            result = client.graph.ro_query("social", "MATCH (n:Person) RETURN n.name")
        """
        raw = self._client.execute_command("GRAPH.RO_QUERY", graph, cypher)
        return _parse_query_result(raw)

    def explain(self, graph: str, cypher: str) -> List[str]:
        """Get the query execution plan without executing.

        Args:
            graph: Graph name.
            cypher: Cypher query string.

        Returns:
            List of plan step descriptions.

        Example::

            plan = client.graph.explain("social", "MATCH (n) RETURN n LIMIT 10")
            for step in plan:
                print(step)
        """
        raw = self._client.execute_command("GRAPH.EXPLAIN", graph, cypher)
        if isinstance(raw, list):
            return [_to_str(x) for x in raw]
        return [_to_str(raw)]

    def profile(self, graph: str, cypher: str) -> List[str]:
        """Execute query and return execution plan with timing stats.

        Args:
            graph: Graph name.
            cypher: Cypher query string.

        Returns:
            List of plan step descriptions with timing.

        Example::

            profile = client.graph.profile("social", "MATCH (n) RETURN n LIMIT 10")
        """
        raw = self._client.execute_command("GRAPH.PROFILE", graph, cypher)
        if isinstance(raw, list):
            return [_to_str(x) for x in raw]
        return [_to_str(raw)]

    def info(self, graph: str) -> Dict[str, Any]:
        """Get graph metadata (node/edge counts, labels, etc.).

        Args:
            graph: Graph name.

        Returns:
            Dictionary of graph metadata.

        Example::

            info = client.graph.info("social")
            print(info.get("num_nodes"), info.get("num_edges"))
        """
        raw = self._client.execute_command("GRAPH.INFO", graph)
        return _parse_info(raw)

    def list_graphs(self) -> List[str]:
        """List all named graphs.

        Returns:
            List of graph names.

        Example::

            graphs = client.graph.list_graphs()
            # ["social", "knowledge"]
        """
        raw = self._client.execute_command("GRAPH.LIST")
        if isinstance(raw, list):
            return [_to_str(x) for x in raw]
        return []

    def vsearch(
        self,
        graph: str,
        start_node_id: int,
        hops: int,
        k: int,
        vector: Union[Sequence[float], bytes],
        *,
        threshold: Optional[float] = None,
        edge_type: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Vector-guided graph search starting from a node.

        Uses GRAPH.VSEARCH to traverse the graph guided by vector similarity.

        Args:
            graph: Graph name.
            start_node_id: Starting node ID.
            hops: Number of graph hops.
            k: Number of results.
            vector: Query vector.
            threshold: Optional distance threshold.
            edge_type: Optional edge type filter.

        Returns:
            List of result dictionaries with node IDs and scores.

        Example::

            results = client.graph.vsearch(
                "knowledge", 1, hops=3, k=5, vector=[0.1, 0.2, ...]
            )
        """
        blob = vector if isinstance(vector, bytes) else encode_vector(vector)
        args: list[Any] = [
            "GRAPH.VSEARCH", graph, str(start_node_id),
            str(hops), str(k), blob,
        ]
        if threshold is not None:
            args.extend(["THRESHOLD", str(threshold)])
        if edge_type:
            args.extend(["TYPE", edge_type])

        raw = self._client.execute_command(*args)
        return _parse_vsearch(raw)


class AsyncGraphCommands:
    """Async variant of GraphCommands for ``redis.asyncio.Redis``.

    All methods are async. See ``GraphCommands`` for documentation.

    Example::

        from moondb import AsyncMoonClient
        client = AsyncMoonClient()
        await client.graph.create("social")
        node_id = await client.graph.add_node("social", "Person", name="Alice")
    """

    def __init__(self, client: Any) -> None:
        self._client = client

    async def create(self, name: str) -> str:
        """Create a named graph (async)."""
        return await self._client.execute_command("GRAPH.CREATE", name)  # type: ignore[no-any-return]

    async def delete(self, name: str) -> str:
        """Delete a named graph (async)."""
        return await self._client.execute_command("GRAPH.DELETE", name)  # type: ignore[no-any-return]

    async def add_node(
        self,
        graph: str,
        label: str,
        *,
        vector_field: Optional[str] = None,
        vector: Optional[Union[Sequence[float], bytes]] = None,
        **properties: str,
    ) -> int:
        """Add a node (async). See ``GraphCommands.add_node``."""
        args: list[Any] = ["GRAPH.ADDNODE", graph, label]
        for k, v in properties.items():
            args.extend([k, v])
        if vector_field and vector is not None:
            blob = vector if isinstance(vector, bytes) else encode_vector(vector)
            args.extend(["VECTOR", vector_field, blob])
        result = await self._client.execute_command(*args)
        return int(result)

    async def add_edge(
        self,
        graph: str,
        src_id: int,
        dst_id: int,
        edge_type: str,
        *,
        weight: Optional[float] = None,
        **properties: str,
    ) -> str:
        """Add an edge (async). See ``GraphCommands.add_edge``."""
        args: list[Any] = ["GRAPH.ADDEDGE", graph, str(src_id), str(dst_id), edge_type]
        for k, v in properties.items():
            args.extend([k, v])
        if weight is not None:
            args.extend(["WEIGHT", str(weight)])
        return await self._client.execute_command(*args)  # type: ignore[no-any-return]

    async def neighbors(
        self,
        graph: str,
        node_id: int,
        *,
        edge_type: Optional[str] = None,
        depth: int = 1,
    ) -> List[GraphNode]:
        """Get neighbors (async). See ``GraphCommands.neighbors``."""
        args: list[Any] = ["GRAPH.NEIGHBORS", graph, str(node_id)]
        if edge_type:
            args.extend(["TYPE", edge_type])
        if depth > 1:
            args.extend(["DEPTH", str(depth)])
        raw = await self._client.execute_command(*args)
        return _parse_neighbors(raw)

    async def query(self, graph: str, cypher: str) -> QueryResult:
        """Execute Cypher query (async). See ``GraphCommands.query``."""
        raw = await self._client.execute_command("GRAPH.QUERY", graph, cypher)
        return _parse_query_result(raw)

    async def ro_query(self, graph: str, cypher: str) -> QueryResult:
        """Execute read-only Cypher query (async). See ``GraphCommands.ro_query``."""
        raw = await self._client.execute_command("GRAPH.RO_QUERY", graph, cypher)
        return _parse_query_result(raw)

    async def explain(self, graph: str, cypher: str) -> List[str]:
        """Get query plan (async). See ``GraphCommands.explain``."""
        raw = await self._client.execute_command("GRAPH.EXPLAIN", graph, cypher)
        if isinstance(raw, list):
            return [_to_str(x) for x in raw]
        return [_to_str(raw)]

    async def profile(self, graph: str, cypher: str) -> List[str]:
        """Execute with profiling (async). See ``GraphCommands.profile``."""
        raw = await self._client.execute_command("GRAPH.PROFILE", graph, cypher)
        if isinstance(raw, list):
            return [_to_str(x) for x in raw]
        return [_to_str(raw)]

    async def info(self, graph: str) -> Dict[str, Any]:
        """Get graph metadata (async). See ``GraphCommands.info``."""
        raw = await self._client.execute_command("GRAPH.INFO", graph)
        return _parse_info(raw)

    async def list_graphs(self) -> List[str]:
        """List all graphs (async). See ``GraphCommands.list_graphs``."""
        raw = await self._client.execute_command("GRAPH.LIST")
        if isinstance(raw, list):
            return [_to_str(x) for x in raw]
        return []

    async def vsearch(
        self,
        graph: str,
        start_node_id: int,
        hops: int,
        k: int,
        vector: Union[Sequence[float], bytes],
        *,
        threshold: Optional[float] = None,
        edge_type: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Vector-guided graph search (async). See ``GraphCommands.vsearch``."""
        blob = vector if isinstance(vector, bytes) else encode_vector(vector)
        args: list[Any] = [
            "GRAPH.VSEARCH", graph, str(start_node_id),
            str(hops), str(k), blob,
        ]
        if threshold is not None:
            args.extend(["THRESHOLD", str(threshold)])
        if edge_type:
            args.extend(["TYPE", edge_type])
        raw = await self._client.execute_command(*args)
        return _parse_vsearch(raw)


# -- Response parsing helpers --


def _to_str(val: Any) -> str:
    """Convert bytes or string to str."""
    if isinstance(val, bytes):
        return val.decode("utf-8", errors="replace")
    return str(val)


def _parse_neighbors(raw: Any) -> List[GraphNode]:
    """Parse GRAPH.NEIGHBORS response.

    Expects a list of [node_id, label, [prop_key, prop_val, ...], ...]
    """
    if not isinstance(raw, list):
        return []

    nodes: List[GraphNode] = []
    for item in raw:
        if isinstance(item, list) and len(item) >= 2:
            node_id = int(item[0])
            label = _to_str(item[1])
            props: Dict[str, str] = {}
            if len(item) > 2 and isinstance(item[2], list):
                j = 0
                while j + 1 < len(item[2]):
                    props[_to_str(item[2][j])] = _to_str(item[2][j + 1])
                    j += 2
            nodes.append(GraphNode(node_id=node_id, label=label, properties=props))

    return nodes


def _parse_query_result(raw: Any) -> QueryResult:
    """Parse GRAPH.QUERY / GRAPH.RO_QUERY response.

    Response format varies -- typically:
    [headers_array, [row1, row2, ...], stats_array]
    """
    if not isinstance(raw, list) or len(raw) < 1:
        return QueryResult()

    headers: List[str] = []
    rows: List[List[Any]] = []
    stats: Dict[str, str] = {}

    # Parse headers (first element)
    if isinstance(raw[0], list):
        headers = [_to_str(h) for h in raw[0]]

    # Parse data rows (second element if present)
    if len(raw) > 1 and isinstance(raw[1], list):
        for row in raw[1]:
            if isinstance(row, list):
                rows.append([_to_str(v) for v in row])
            else:
                rows.append([_to_str(row)])

    # Parse stats (last element if it's a list of strings)
    if len(raw) > 2 and isinstance(raw[-1], list):
        for stat in raw[-1]:
            s = _to_str(stat)
            if ":" in s:
                k, _, v = s.partition(":")
                stats[k.strip()] = v.strip()

    return QueryResult(headers=headers, rows=rows, stats=stats)


def _parse_info(raw: Any) -> Dict[str, Any]:
    """Parse GRAPH.INFO response (flat key-value list)."""
    if not isinstance(raw, list):
        return {}

    info: Dict[str, Any] = {}
    i = 0
    while i + 1 < len(raw):
        key = _to_str(raw[i])
        val = raw[i + 1]
        if isinstance(val, bytes):
            info[key] = val.decode("utf-8", errors="replace")
        else:
            info[key] = val
        i += 2

    return info


def _parse_vsearch(raw: Any) -> List[Dict[str, Any]]:
    """Parse GRAPH.VSEARCH response."""
    if not isinstance(raw, list):
        return []

    results: List[Dict[str, Any]] = []
    for item in raw:
        if isinstance(item, list) and len(item) >= 2:
            results.append({
                "node_id": int(item[0]) if not isinstance(item[0], int) else item[0],
                "score": float(_to_str(item[1])) if len(item) > 1 else 0.0,
            })
    return results

File: sdk/rust/src/graph.rs
use std::collections::HashMap;
use crate::error::Result;
use crate::types::{GraphEdge, GraphNode, QueryResult};
use crate::util::{parse_query_result, value_to_i64, value_to_string};

/// Graph engine sub-client (`GRAPH.*` commands).
///
/// Obtain via [`MoonClient::graph`](crate::MoonClient::graph).
///
/// # Example
/// ```no_run
/// use moon_client::MoonClient;
///
/// #[tokio::main]
/// async fn main() -> moon_client::Result<()> {
///     let client = MoonClient::connect("redis://127.0.0.1:6399").await?;
///     let mut g = client.graph();
///     g.create("social").await?;
///     let alice = g.add_node("social", "Person", &[("name", "Alice"), ("age", "30")]).await?;
///     let bob = g.add_node("social", "Person", &[("name", "Bob")]).await?;
///     g.add_edge("social", alice, bob, "KNOWS", 1.0, &[]).await?;
///     let result = g.query("social", "MATCH (p:Person) RETURN p.name").await?;
///     for row in result.rows {
///         println!("{:?}", row);
///     }
///     Ok(())
/// }
/// ```
pub struct GraphClient {
    pub(crate) conn: redis::aio::MultiplexedConnection,
}

impl GraphClient {
    // ── Graph lifecycle ──────────────────────────────────────────────────────

    /// Create a named graph (`GRAPH.CREATE`).
    pub async fn create(&mut self, name: &str) -> Result<()> {
        redis::cmd("GRAPH.CREATE")
            .arg(name)
            .query_async::<()>(&mut self.conn)
            .await?;
        Ok(())
    }

    /// Drop a graph and all its data (`GRAPH.DELETE` or `GRAPH.DROP`).
    pub async fn delete(&mut self, name: &str) -> Result<()> {
        redis::cmd("GRAPH.DELETE")
            .arg(name)
            .query_async::<()>(&mut self.conn)
            .await?;
        Ok(())
    }

    /// List all graphs (`GRAPH.LIST`).
    pub async fn list(&mut self) -> Result<Vec<String>> {
        Ok(redis::cmd("GRAPH.LIST").query_async(&mut self.conn).await?)
    }

    /// Get graph statistics (`GRAPH.INFO`).
    pub async fn info(&mut self, name: &str) -> Result<redis::Value> {
        Ok(redis::cmd("GRAPH.INFO").arg(name).query_async(&mut self.conn).await?)
    }

    // ── Nodes ────────────────────────────────────────────────────────────────

    /// Add a node and return its server-assigned numeric ID (`GRAPH.ADDNODE`).
    ///
    /// `properties` is a slice of `(field_name, field_value)` pairs.
    pub async fn add_node(
        &mut self,
        graph: &str,
        label: &str,
        properties: &[(&str, &str)],
    ) -> Result<i64> {
        let mut cmd = redis::cmd("GRAPH.ADDNODE");
        cmd.arg(graph).arg(label);
        for (k, v) in properties {
            cmd.arg(*k).arg(*v);
        }
        let raw: redis::Value = cmd.query_async(&mut self.conn).await?;
        Ok(value_to_i64(&raw).unwrap_or(0))
    }

    /// Retrieve a node by ID (`GRAPH.GETNODE`). Returns `None` if not found.
    pub async fn get_node(&mut self, graph: &str, node_id: i64) -> Result<Option<GraphNode>> {
        let raw: redis::Value = redis::cmd("GRAPH.GETNODE")
            .arg(graph)
            .arg(node_id)
            .query_async(&mut self.conn)
            .await?;
        Ok(parse_node_value(raw))
    }

    /// Delete a node by ID (`GRAPH.DELNODE`).
    pub async fn del_node(&mut self, graph: &str, node_id: i64) -> Result<()> {
        redis::cmd("GRAPH.DELNODE")
            .arg(graph)
            .arg(node_id)
            .query_async::<()>(&mut self.conn)
            .await?;
        Ok(())
    }

    // ── Edges ────────────────────────────────────────────────────────────────

    /// Add a directed edge between two nodes (`GRAPH.ADDEDGE`).
    pub async fn add_edge(
        &mut self,
        graph: &str,
        src_id: i64,
        dst_id: i64,
        edge_type: &str,
        weight: f64,
        properties: &[(&str, &str)],
    ) -> Result<()> {
        let mut cmd = redis::cmd("GRAPH.ADDEDGE");
        cmd.arg(graph).arg(src_id).arg(dst_id).arg(edge_type).arg("WEIGHT").arg(weight);
        for (k, v) in properties {
            cmd.arg(*k).arg(*v);
        }
        cmd.query_async::<()>(&mut self.conn).await?;
        Ok(())
    }

    /// Retrieve an edge by source and destination IDs (`GRAPH.GETEDGE`).
    pub async fn get_edge(&mut self, graph: &str, src_id: i64, dst_id: i64) -> Result<Option<GraphEdge>> {
        let raw: redis::Value = redis::cmd("GRAPH.GETEDGE")
            .arg(graph)
            .arg(src_id)
            .arg(dst_id)
            .query_async(&mut self.conn)
            .await?;
        Ok(parse_edge_value(raw))
    }

    /// Delete an edge (`GRAPH.DELEDGE`).
    pub async fn del_edge(&mut self, graph: &str, src_id: i64, dst_id: i64) -> Result<()> {
        redis::cmd("GRAPH.DELEDGE")
            .arg(graph)
            .arg(src_id)
            .arg(dst_id)
            .query_async::<()>(&mut self.conn)
            .await?;
        Ok(())
    }

    // ── Traversal ────────────────────────────────────────────────────────────

    /// Get neighbors of a node (`GRAPH.NEIGHBORS`).
    ///
    /// Returns a list of `(src_id, dst_id, edge_type, weight)` tuples.
    pub async fn neighbors(
        &mut self,
        graph: &str,
        node_id: i64,
        direction: NeighborDirection,
    ) -> Result<Vec<(i64, i64, String, f64)>> {
        let _ = direction; // server does not filter by direction; kept for API forwards-compat
        let raw: redis::Value = redis::cmd("GRAPH.NEIGHBORS")
            .arg(graph)
            .arg(node_id)
            .query_async(&mut self.conn)
            .await?;
        Ok(crate::util::parse_neighbors(raw))
    }

    /// Get degree of a node (`GRAPH.DEGREE`).
    pub async fn degree(&mut self, graph: &str, node_id: i64) -> Result<i64> {
        Ok(redis::cmd("GRAPH.DEGREE")
            .arg(graph)
            .arg(node_id)
            .query_async(&mut self.conn)
            .await?)
    }

    // ── Cypher queries ───────────────────────────────────────────────────────

    /// Execute a read-write Cypher query (`GRAPH.QUERY`).
    pub async fn query(&mut self, graph: &str, cypher: &str) -> Result<QueryResult> {
        let raw: redis::Value = redis::cmd("GRAPH.QUERY")
            .arg(graph)
            .arg(cypher)
            .query_async(&mut self.conn)
            .await?;
        Ok(parse_query_result(raw))
    }

    /// Execute a read-only Cypher query (`GRAPH.RO_QUERY`).
    pub async fn ro_query(&mut self, graph: &str, cypher: &str) -> Result<QueryResult> {
        let raw: redis::Value = redis::cmd("GRAPH.RO_QUERY")
            .arg(graph)
            .arg(cypher)
            .query_async(&mut self.conn)
            .await?;
        Ok(parse_query_result(raw))
    }

    /// Temporal Cypher query at a specific wall-clock timestamp (`GRAPH.QUERY ... VALID_AT <ms>`).
    pub async fn query_at(&mut self, graph: &str, cypher: &str, valid_at_ms: i64) -> Result<QueryResult> {
        let raw: redis::Value = redis::cmd("GRAPH.QUERY")
            .arg(graph)
            .arg(cypher)
            .arg("VALID_AT")
            .arg(valid_at_ms)
            .query_async(&mut self.conn)
            .await?;
        Ok(parse_query_result(raw))
    }

    /// Explain a Cypher query execution plan (`GRAPH.EXPLAIN`).
    pub async fn explain(&mut self, graph: &str, cypher: &str) -> Result<Vec<String>> {
        Ok(redis::cmd("GRAPH.EXPLAIN")
            .arg(graph)
            .arg(cypher)
            .query_async(&mut self.conn)
            .await?)
    }

    /// Profile a Cypher query with execution stats (`GRAPH.PROFILE`).
    pub async fn profile(&mut self, graph: &str, cypher: &str) -> Result<redis::Value> {
        Ok(redis::cmd("GRAPH.PROFILE")
            .arg(graph)
            .arg(cypher)
            .query_async(&mut self.conn)
            .await?)
    }

    /// Vector-similarity graph search (`GRAPH.VSEARCH`).
    ///
    /// Traverses up to `hops` from `start_node_id`, scores candidates by cosine
    /// similarity to `query_vec`, and returns the top-`k` results.
    pub async fn vsearch(
        &mut self,
        graph: &str,
        start_node_id: i64,
        hops: usize,
        k: usize,
        query_vec: &[f32],
    ) -> Result<redis::Value> {
        let blob = crate::types::encode_vector(query_vec);
        Ok(redis::cmd("GRAPH.VSEARCH")
            .arg(graph)
            .arg(start_node_id)
            .arg(hops)
            .arg(k)
            .arg(&blob)
            .query_async(&mut self.conn)
            .await?)
    }
}

// ── Edge/node response parsers ───────────────────────────────────────────────

fn parse_node_value(raw: redis::Value) -> Option<GraphNode> {
    let arr = match raw {
        redis::Value::Array(a) => a,
        redis::Value::Nil => return None,
        _ => return None,
    };
    if arr.is_empty() {
        return None;
    }
    let kv = flat_kv(&arr);
    let node_id = kv.get("node_id")?.parse().ok()?;
    let label = kv.get("label").cloned().unwrap_or_default();
    let mut properties = HashMap::new();
    for (k, v) in kv {
        if !matches!(k.as_str(), "node_id" | "label") {
            properties.insert(k, v);
        }
    }
    Some(GraphNode { node_id, label, properties })
}

fn parse_edge_value(raw: redis::Value) -> Option<GraphEdge> {
    let arr = match raw {
        redis::Value::Array(a) => a,
        redis::Value::Nil => return None,
        _ => return None,
    };
    if arr.is_empty() {
        return None;
    }
    let kv = flat_kv(&arr);
    let src_id = kv.get("src_id")?.parse().ok()?;
    let dst_id = kv.get("dst_id")?.parse().ok()?;
    let edge_type = kv.get("edge_type").cloned().unwrap_or_default();
    let weight: f64 = kv.get("weight").and_then(|v| v.parse().ok()).unwrap_or(1.0);
    let mut properties = HashMap::new();
    for (k, v) in kv {
        if !matches!(k.as_str(), "src_id" | "dst_id" | "edge_type" | "weight") {
            properties.insert(k, v);
        }
    }
    Some(GraphEdge { src_id, dst_id, edge_type, weight, properties })
}

fn flat_kv(arr: &[redis::Value]) -> HashMap<String, String> {
    let mut m = HashMap::new();
    let mut i = 0;
    while i + 1 < arr.len() {
        let k = value_to_string(&arr[i]);
        let v = value_to_string(&arr[i + 1]);
        if !k.is_empty() {
            m.insert(k, v);
        }
        i += 2;
    }
    m
}

/// Direction filter for `GRAPH.NEIGHBORS`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NeighborDirection {
    Both,
    Out,
    In,
}

impl NeighborDirection {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Both => "BOTH",
            Self::Out => "OUT",
            Self::In => "IN",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn neighbor_direction_strings() {
        assert_eq!(NeighborDirection::Both.as_str(), "BOTH");
        assert_eq!(NeighborDirection::Out.as_str(), "OUT");
        assert_eq!(NeighborDirection::In.as_str(), "IN");
    }

    #[test]
    fn parse_empty_node_value() {
        assert!(parse_node_value(redis::Value::Nil).is_none());
        assert!(parse_node_value(redis::Value::Array(vec![])).is_none());
    }
}
